scipy/special/_multiufuncs.py

```diff
@@ -0,0 +1,640 @@
+import numpy as np
+
+from ._input_validation import _nonneg_int_or_fail
+
+from ._special_ufuncs import (legendre_p, assoc_legendre_p,
+                              sph_legendre_p, sph_harm_y)
+from ._gufuncs import (legendre_p_all, assoc_legendre_p_all,
+                       sph_legendre_p_all, sph_harm_y_all)
+
+__all__ = [
+    "assoc_legendre_p",
+    "assoc_legendre_p_all",
+    "legendre_p",
+    "legendre_p_all",
+    "sph_harm_y",
+    "sph_harm_y_all",
+    "sph_legendre_p",
+    "sph_legendre_p_all",
+]
+
+
+class MultiUFunc:
+    def __init__(self, ufunc_or_ufuncs, doc=None, *,
+                 force_complex_output=False, **default_kwargs):
+        if not isinstance(ufunc_or_ufuncs, np.ufunc):
+            if isinstance(ufunc_or_ufuncs, collections.abc.Mapping):
+                ufuncs_iter = ufunc_or_ufuncs.values()
+            elif isinstance(ufunc_or_ufuncs, collections.abc.Iterable):
+                ufuncs_iter = ufunc_or_ufuncs
+            else:
+                raise ValueError("ufunc_or_ufuncs should be a ufunc or a"
+                                 " ufunc collection")
+
+            # Perform input validation to ensure all ufuncs in ufuncs are
+            # actually ufuncs and all take the same input types.
+            seen_input_types = set()
+            for ufunc in ufuncs_iter:
+                if not isinstance(ufunc, np.ufunc):
+                    raise ValueError("All ufuncs must have"
+                                     f" type `numpy.ufunc`. Received {ufunc_or_ufuncs}")
+                seen_input_types.add(frozenset(x.split("->")[0] for x in ufunc.types))
+            if len(seen_input_types) > 1:
+                raise ValueError("All ufuncs must take the same input types.")
+
+        self._ufunc_or_ufuncs = ufunc_or_ufuncs
+        self.__doc = doc
+        self.__force_complex_output = force_complex_output
+        self._default_kwargs = default_kwargs
+        self.resolve_out_shapes = None
+        self.key = None
+        self.ufunc_default_args = lambda *args, **kwargs: ()
+        self.ufunc_default_kwargs = lambda *args, **kwargs: {}
+
+    @property
+    def __doc__(self):
+        return self.__doc
+
+    @property
+    def force_complex_output(self):
+        return self.__force_complex_output
+
+    def override_key(self, func):
+        """Set `key` method by decorating a function.
+        """
+        self.key = func
+
+    def override_ufunc_default_args(self, func):
+        self.ufunc_default_args = func
+
+    def override_ufunc_default_kwargs(self, func):
+        self.ufunc_default_kwargs = func
+
+    def override_resolve_out_shapes(self, func):
+        """Set `resolve_out_shapes` method by decorating a function."""
+        if func.__doc__ is None:
+            func.__doc__ = \
+                """Resolve to output shapes based on relevant inputs."""
+        func.__name__ = "resolve_out_shapes"
+        self.resolve_out_shapes = func
+
+    def resolve_ufunc(self, **kwargs):
```

Comment:
Before your latest round of commits, I had made it so that `resolve_ufunc` had an accurate signature. Is this just an inadvertent change? I'd prefer to keep the accurate signatures if possible. Right now on your branch we have:

```
In [10]: assoc_legendre_p_all.resolve_ufunc?
Signature: assoc_legendre_p_all.resolve_ufunc(**kwargs)
Docstring: Resolve to a ufunc based on keyword arguments.
File:      ~/scipy/build-install/lib/python3.12/site-packages/scipy/special/_multiufuncs.py
Type:      method
```

but on an earlier commit this was

```
In [2]: assoc_legendre_p_all.resolve_ufunc?
Signature: assoc_legendre_p_all.resolve_ufunc(norm=False, diff_n=0)
Docstring: Resolve to ufunc dispatched to for given keyword arguments.
File:      ~/scipy/build-install/lib/python3.12/site-packages/scipy/special/_multiufuncs.py
Type:      function
```

Reply:
We can do this is required. But what changed is that `resolve_ufunc` is no longer something the user implements, it just calls the `key` function. It itself never changes.

To do what you are suggesting would require monkeypatching it, but we have no reason to do that anymore. I say we leave it as is for simplicity. Again, though, happy to do what you are suggesting for any other relevant functions.

Reply:
> What changed is that `resolve_ufunc` is no longer something the user implements, it just calls the `key` function. It itself never changes.

I think there's a distinction we need to make clear. The users of `MultiUFunc` are the people who import the special functions and use them in their code. Users were never required to implement `resolve_ufunc`, what we're discussing is the internal API for setting up `MultiUFunc`s for the user. Monkeypatching is no longer required to get the functionality to work correctly, but having a proper signature is very helpful for the user, especially because some of a `MultiUFunc`s kwargs are used to give ufuncs default args and some are used to choose which ufunc to dispatch to. I think our main consideration in designing `MultiUFunc` should be the user facing API, not how we set things up internally.

Reply:
Ok, we've come to a compromise that `resolve_ufunc` shall be private for now. I have turned it into `_resolve_ufunc`.